### ISSUE TITLE
syslog-ng is stuck during loading mongodb plugin

### ISSUE THREAD

**Issue by mitzkia** [balabit/syslog-ng#1350]
# syslog-ng

## Version of syslog-ng

```
OSE master: 3.9.1
builded from source
last commit: 4b0b39971949dbc7844d61a232b2b6b370e01510
```

## Platform

```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.10
Release:	16.10
Codename:	yakkety

$ uname -a
Linux tristram 4.8.0-37-generic #39-Ubuntu SMP Thu Jan 26 02:27:07 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

# Issue

## Failure

I know that very well that this is (also) not a usual working mode for syslog-ng.
After the reproduction /dev/log can not be used (sudo will not work).

Reproduction steps, 

Advice:
Before the reproduction create a new shell env. with root privileges to kill out syslog-ng when it stucked (pkill -9 syslog-ng).

1, start syslog-ng with the following config:

```
sudo sbin/syslog-ng -Fedvt -f syslog-ng.conf --no-caps --enable-core
```

```
@version: 3.9
@include "scl.conf"
source s_system {system();};
destination d_file {file("/tmp/output");};

log{
        source(s_system);
        destination(d_file);
        #flags(flow-control);
};
```

2, Use the following command line to trigger the stuck for syslog-ng (maybe syslog-ng will not stuck for the first time, but for the second or third time when the script is used)

```
$ for i in {1..500} ; do echo "$i" ; sudo pkill -1 syslog-ng ; done
```

I have got the following bt when it stucked:

bt full: 
[gdb_core.txt](https://github.com/balabit/syslog-ng/files/784171/gdb_core.txt)

```
(gdb) bt
#0  0x00007f93f2c2310f in __libc_send (fd=21, buf=0x560aded26600, n=58, flags=flags@entry=16384) at ../sysdeps/unix/sysv/linux/x86_64/send.c:26
#1  0x00007f93f2c1b1fa in __GI___vsyslog_chk (pri=<optimized out>, flag=<optimized out>, fmt=0x7f93eec7e90e "%s", ap=0x7ffd0ee63c18) at ../misc/syslog.c:279
#2  0x00007f93f2c1b462 in __syslog_chk (pri=<optimized out>, flag=<optimized out>, fmt=<optimized out>) at ../misc/syslog.c:129
#3  0x00007f93eec71623 in ?? () from /usr/lib/x86_64-linux-gnu/libsasl2.so.2
#4  0x00007f93eec731ff in ?? () from /usr/lib/x86_64-linux-gnu/libsasl2.so.2
#5  0x00007f93ee24579a in ?? () from /usr/lib/x86_64-linux-gnu/sasl2/libdigestmd5.so
#6  0x00007f93eec6fa69 in ?? () from /usr/lib/x86_64-linux-gnu/libsasl2.so.2
#7  0x00007f93eec6fbaa in sasl_client_done () from /usr/lib/x86_64-linux-gnu/libsasl2.so.2
#8  0x00007f93eeed3ddb in _mongoc_do_cleanup () at ../../../../modules/afmongodb/mongo-c-driver/src/mongoc/mongoc-init.c:130
#9  0x00007f93f2eefc19 in __pthread_once_slow (once_control=0x7f93ef121c44 <once>, init_routine=0x7f93eeed3dcd <_mongoc_do_cleanup>) at pthread_once.c:116
#10 0x00007f93eeed3dfa in mongoc_cleanup () at ../../../../modules/afmongodb/mongo-c-driver/src/mongoc/mongoc-init.c:150
#11 0x00007f93eeed3e12 in _mongoc_init_dtor () at ../../../../modules/afmongodb/mongo-c-driver/src/mongoc/mongoc-init.c:169
#12 0x00007f93f370a479 in _dl_close_worker (map=map@entry=0x560aded23de0, force=force@entry=false) at dl-close.c:286
#13 0x00007f93f370b27a in _dl_close_worker (force=false, map=0x560aded23de0) at dl-close.c:125
#14 _dl_close (_map=0x560aded23de0) at dl-close.c:822
#15 0x00007f93f3704674 in _dl_catch_error (objname=0x560adecb0bf0, errstring=0x560adecb0bf8, mallocedp=0x560adecb0be8, operate=0x7f93f15dcfb0 <dlclose_doit>, args=0x560aded23de0) at dl-error.c:187
#16 0x00007f93f15dd571 in _dlerror_run (operate=operate@entry=0x7f93f15dcfb0 <dlclose_doit>, args=0x560aded23de0) at dlerror.c:163
#17 0x00007f93f15dcfdf in __dlclose (handle=<optimized out>) at dlclose.c:46
#18 0x00007f93f270f3c2 in g_module_close () from /usr/lib/x86_64-linux-gnu/libgmodule-2.0.so.0
#19 0x00007f93f3461f08 in plugin_load_candidate_modules (cfg=0x560aded44740) at ../lib/plugin.c:456
#20 0x00007f93f344159f in cfg_load_candidate_modules (self=0x560aded44740) at ../lib/cfg.c:384
#21 0x00007f93f34442f6 in cfg_lexer_lex (self=0x560adece4880, yylval=0x7ffd0ee645a0, yylloc=0x7ffd0ee645c0) at ../lib/cfg-lexer.c:938
#22 0x00007f93f3445861 in main_lex (yylval=0x7ffd0ee645a0, yylloc=0x7ffd0ee645c0, lexer=0x560adece4880) at ../lib/cfg-parser.c:182
#23 0x00007f93f347466d in main_parse (lexer=0x560adece4880, dummy=0x7ffd0ee66988, arg=0x0) at lib/cfg-grammar.c:3078
#24 0x00007f93f344075b in cfg_parser_parse (self=0x7f93f36eff60 <main_parser>, lexer=0x560adece4880, instance=0x7ffd0ee66988, arg=0x0) at ../lib/cfg-parser.h:83
#25 0x00007f93f3441501 in cfg_run_parser (self=0x560aded44740, lexer=0x560adece4880, parser=0x7f93f36eff60 <main_parser>, result=0x7ffd0ee66988, arg=0x0) at ../lib/cfg.c:366
#26 0x00007f93f3441744 in cfg_read_config (self=0x560aded44740, fname=0x560adec80430 "/home/micek/syslog_ng_crashes/syslog-ng-compact22345678912345678/syslog-ng-compact22345678912345678.conf", syntax_only=0, 
    preprocess_into=0x0) at ../lib/cfg.c:438
#27 0x00007f93f345b997 in main_loop_reload_config_initiate (user_data=0x7f93f36f2a80 <main_loop>) at ../lib/mainloop.c:277
#28 0x00007f93f345bb9e in sig_hup_handler (user_data=0x7f93f36f2a80 <main_loop>) at ../lib/mainloop.c:345
#29 0x00007f93f34ab9e9 in iv_signal_event (_this=0x7f93f36f2a88 <main_loop+8>) at ../../../../lib/ivykis/src/iv_signal.c:170
#30 0x00007f93f34a9b8f in iv_event_raw_got_event (_this=0x7f93f36f2ac8 <main_loop+72>) at ../../../../lib/ivykis/src/iv_event_raw_posix.c:89
#31 0x00007f93f34aa2fc in iv_fd_poll_and_run (st=0x560adec77e80, to=0x7ffd0ee66fa0) at ../../../../lib/ivykis/src/iv_fd.c:163
#32 0x00007f93f34ab2d6 in iv_main () at ../../../../lib/ivykis/src/iv_main_posix.c:117
#33 0x00007f93f345c0f9 in main_loop_run (self=0x7f93f36f2a80 <main_loop>) at ../lib/mainloop.c:519
#34 0x0000560adea0905b in main (argc=1, argv=0x7ffd0ee67118) at ../syslog-ng/main.c:300
```

**Comment by mitzkia** [comment]
Can be fixed by: #1765

**Comment by MrAnno** [comment]
Fixed in #1765.
